File: spec/features/feed_management_spec.rb
require 'spec_helper'

feature 'Feed Management' do
  scenario 'User subscribes to a feed' do
    create_user name: 'john', password: 'secret'
    login name: 'john', password: 'secret'

    click_on 'Feeds'

    fill_in 'Feed URL', with: 'http://localhost:1234'
    click_on 'Subscribe'

    within '.feeds > ul' do
      expect(page).to have_content 'http://localhost:1234'
    end
  end

  scenario 'User imports OPML file' do
    create_user name: 'john', password: 'secret'
    login name: 'john', password: 'secret'

    click_on 'Feeds'

    attach_file 'OPML File', File.expand_path('spec/fixtures/opml.xml')
    click_on 'Import'

    accept_alert

    within '.feeds > ul' do
      expect(page).to have_content 'http://localhost/rss'
      expect(page).to have_content 'http://localhost/other/rss'
    end
  end
end

File: e2e/tests/feed-management.spec.ts
import { test, expect } from '@playwright/test';
import { resetDatabase, createUser } from '../fixtures/database';
import { login } from '../fixtures/test-helpers';
import path from 'path';

test.describe('Feed Management', () => {
  test.beforeEach(async () => {
    await resetDatabase();
  });

  test('user subscribes to a feed', async ({ page }) => {
    // Create user and login
    await createUser({ name: 'john', password: 'secret' });
    await login(page, 'john', 'secret');

    // Navigate to feeds page
    await page.getByRole('link', { name: 'Feeds' }).click();

    // Wait for feeds page to load
    await expect(page.getByLabel('Feed URL')).toBeVisible();

    // Subscribe to a feed
    await page.getByLabel('Feed URL').fill('http://localhost:1234');
    await page.getByRole('button', { name: 'Subscribe' }).click();

    // Verify feed appears in the list
    await expect(page.locator('.feeds > ul')).toContainText('http://localhost:1234');
  });

  test('user imports OPML file', async ({ page }) => {
    // Create user and login
    await createUser({ name: 'john', password: 'secret' });
    await login(page, 'john', 'secret');

    // Navigate to feeds page
    await page.getByRole('link', { name: 'Feeds' }).click();

    // Wait for feeds page to load
    await expect(page.locator('[type=file]')).toBeVisible();

    // Upload OPML file
    const opmlPath = path.join(__dirname, '../../test/testdata/opml.xml');
    await page.locator('[type=file]').setInputFiles(opmlPath);

    // Handle dialog and click import
    page.once('dialog', dialog => dialog.accept());
    await page.getByRole('button', { name: 'Import' }).click();

    // Verify feeds from OPML appear in the list
    await expect(page.locator('.feeds > ul')).toContainText('http://localhost/rss');
    await expect(page.locator('.feeds > ul')).toContainText('http://localhost/other/rss');
  });
});
